File: 2020/day3/Part2.java
/*
 * Advent of Code 2020 - Day 3 - Part 2
 * Copyright (C) 2020  Michael Federczuk
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import java.io.FileInputStream;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class Part2 {
	enum Field {
		TREE,
		NOT_TREE
	};

	static class Slope {
		public int right;
		public int down;

		public Slope(int right, int down) {
			this.right = right;
			this.down = down;
		}
	}

	public static void main(String[] args) throws IOException {
		String inputFile = "example_input";
		if(args.length >= 1) inputFile = args[0];

		List<List<Field>> map = new ArrayList<List<Field>>();

		var inputFileScanner = new Scanner(new FileInputStream(inputFile));

		while(inputFileScanner.hasNext()) {
			var line = inputFileScanner.nextLine();

			var row = new ArrayList<Field>();
			map.add(row);

			for(char ch : line.toCharArray()) {
				var field = (ch == '#' ? Field.TREE : Field.NOT_TREE);
				row.add(field);
			}
		}

		inputFileScanner.close();

		List<Slope> slopes = new ArrayList<Slope>();
		slopes.add(new Slope(1, 1));
		slopes.add(new Slope(3, 1));
		slopes.add(new Slope(5, 1));
		slopes.add(new Slope(7, 1));
		slopes.add(new Slope(1, 2));

		long treeHitCount = 1;

		for(var slope : slopes) {
			int slopeTreeHitCount = 0;

			for(int i = 0, rowIndex; (rowIndex = (i * slope.down)) < map.size(); ++i) {
				var row = map.get(rowIndex);
				int columnIndex = (i * slope.right) % row.size();

				var field = row.get(columnIndex);
				if(field == Field.TREE) ++slopeTreeHitCount;
			}

			treeHitCount *= slopeTreeHitCount;
		}

		System.out.println(treeHitCount);
	}
}

File: 2020/day4/part2.ts
/*
 * Advent of Code 2020 - Day 4 - Part 2
 * Copyright (C) 2020  Michael Federczuk
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import { readFileSync } from "fs";
import { MapLike } from "typescript";

let inputFilePath = "part2_example_input_valid";
if(process.argv.length >= 3) inputFilePath = process.argv[2];

const inputFile = readFileSync(inputFilePath).toString();

const passports = inputFile.split("\n\n").map((passportStr) => {
	return passportStr.trim().split(/\s+/).reduce((passportObj, field) => {
		const [key, value] = field.split(":");
		passportObj[key] = value;
		return passportObj;
	}, {} as MapLike<string>);
});

const validEyeColors = [
	"amb",
	"blu",
	"brn",
	"gry",
	"grn",
	"hzl",
	"oth"
];

const requiredFieldValidators: [string, (value: string) => boolean][] = [
	["byr", (value) => {
		if(value.match(/^[0-9]{4}$/) == null) return false;
		const byr = Number(value);
		return (byr >= 1920 && byr <= 2002);
	}],
	["iyr", (value) => {
		if(value.match(/^[0-9]{4}$/) == null) return false;
		const iyr = Number(value);
		return (iyr >= 2010 && iyr <= 2020);
	}],
	["eyr", (value) => {
		if(value.match(/^[0-9]{4}$/) == null) return false;
		const eyr = Number(value);
		return (eyr >= 2020 && eyr <= 2030);
	}],
	["hgt", (value) => {
		const match = value.match(/^(?<hgtStr>[0-9]+)(?<unit>cm|in)$/);
		if(match === null) return false;

		const { hgtStr, unit } = match.groups!!;
		const hgt = Number(hgtStr);

		switch(unit) {
			case("cm"): return (hgt >= 150 && hgt <= 193);
			case("in"): return (hgt >= 59 && hgt <= 76);
		}
		throw undefined;
	}],
	["hcl", (value) => (value.match(/^#[0-9a-f]{6}$/) != null)],
	["ecl", (value) => (validEyeColors.includes(value))],
	["pid", (value) => (value.match(/^[0-9]{9}$/) != null)]
];

const numOfValidPassports = passports.reduce((numOfValidPassports, passport) => {
	const passportValid = requiredFieldValidators.every(([requiredFieldKey, requiredFieldValidator]) => {
		const fieldValue = passport[requiredFieldKey];
		return (fieldValue !== undefined && requiredFieldValidator(fieldValue));
	});

	if(passportValid) ++numOfValidPassports;

	return numOfValidPassports;
}, 0);

console.log(numOfValidPassports);
